reporter/sbom/spdx.go

```diff
@@ -0,0 +1,587 @@
+				packages = append(packages, &pack)
+				relationships = append(relationships, makeSPDXRelationShip(ghpkgs[0].PackageSPDXIdentifier, pack.PackageSPDXIdentifier, relationShipDepensOn))
+			}
+		}
+	}
+
+	if wppkgs := wppkgToSPDXPackages(result.WordPressPackages, packageToURLMap, result.ReportedAt); len(wppkgs) > 0 {
+		packages = append(packages, &wppkgs[0])
+		relationships = append(relationships, makeSPDXRelationShip(root.PackageSPDXIdentifier, wppkgs[0].PackageSPDXIdentifier, relationShipContains))
+		for _, pack := range wppkgs[1:] {
+			packages = append(packages, &pack)
+			relationships = append(relationships, makeSPDXRelationShip(wppkgs[0].PackageSPDXIdentifier, pack.PackageSPDXIdentifier, relationShipDepensOn))
+		}
+	}
+
+	return packages, relationships
+}
+
+func ospkgToSPDXPackages(r models.ScanResult, packageToURLMap map[string][]string) []spdx.Package {
+	if r.Family == "" || len(r.Packages) == 0 {
+		return nil
+	}
+
+	type srcpkg struct {
+		name    string
+		version string
+		arch    string
+	}
+	binToSrc := map[string]srcpkg{}
+	for _, pack := range r.SrcPackages {
+		for _, binpkg := range pack.BinaryNames {
+			binToSrc[binpkg] = srcpkg{
+				name:    pack.Name,
+				version: pack.Version,
+				arch:    pack.Arch,
+			}
+		}
+	}
+
+	packages := make([]spdx.Package, 0, len(r.Packages))
+	for _, pack := range r.Packages {
+		var annotations []spdx.Annotation
+		if sp, ok := binToSrc[pack.Name]; ok {
+			annotations = appendAnnotation(annotations, "SrcName", sp.name, r.ReportedAt)
+			annotations = appendAnnotation(annotations, "SrcVersion", sp.version, r.ReportedAt)
+			annotations = appendAnnotation(annotations, "SrcArch", sp.arch, r.ReportedAt)
+		}
+
+		var externalRefs []*spdx.PackageExternalReference
+		purl := osPkgToPURL(r.Family, r.Release, pack)
+		externalRefs = appendExternalRefs(externalRefs, categoryPackageManager, packageManagerPURL, purl.String())
+
+		if urls, exists := packageToURLMap[pack.Name]; exists {
+			for _, url := range urls {
+				externalRefs = appendExternalRefs(externalRefs, categorySecurity, securityAdvisory, url)
+			}
+		}
+
+		spdxPackage := spdx.Package{
+			PackageSPDXIdentifier:     calculateSDPXIDentifier(elementPackage),
+			PackageName:               pack.Name,
+			PackageVersion:            pack.Version,
+			PackageDownloadLocation:   noneField,
+			Annotations:               annotations,
+			PackageExternalReferences: externalRefs,
+		}
+		packages = append(packages, spdxPackage)
+	}
+	return packages
+}
+
+func cpeToSPDXPackages(r models.ScanResult, packageToURLMap map[string][]string) []spdx.Package {
+	cpes := map[string]struct{}{}
+	for _, cve := range r.ScannedCves {
+		for _, cpe := range cve.CpeURIs {
+			cpes[cpe] = struct{}{}
+		}
+	}
+
+	if len(cpes) == 0 {
+		return nil
+	}
+
+	packages := make([]spdx.Package, 0, 1+len(cpes))
+
+	packages = append(packages, spdx.Package{
+		PackageSPDXIdentifier:   calculateSDPXIDentifier(elementPackage),
+		PackageName:             "CPEs",
```

Comment:
I don't know SPDX spec or implementations... the word "CPEs" makes me feel it is like type or class thingy.  If it is used like in this way in some tools, it's OK.  Hmm....

Reply:
My miunderstanding, sorry. Under the "CPEs", entries for indivisual cpes exist. Thank you for the explanation!